guava/src/com/google/common/io/BaseEncoding.java

```diff
@@ -468,19 +477,21 @@ public static BaseEncoding base16() {
       this.validPadding = validPadding;
     }
 
-    char encode(int bits) {
+    char encode(@IndexFor("this.chars") int bits) {
       return chars[bits];
     }
 
+    @SuppressWarnings("array.access.unsafe") // The index used is smaller than charsPerChunk, which is the length of validPadding
```

Comment:
This should verify if you annotate `index` as `@NonNegative` - modulus is negative if its input is negative.

Reply:
Annotating `index` with `@NonNegative` works for the lower bound error, but the higher bound error persists. `@LessThan(charsPerChunk)` or `@IndexFor(this.validPadding)` on `index` are not accurate, so I can't see a way around this.